src/js/proxified-containers.js

```diff
@@ -0,0 +1,148 @@
+      browser.storage.local.get("proxifiedContainersKey").then((results) => {
+        // Steps to test:
+        // 1. Is result empty? If so we must inform the caller to intialize proxifiedContainersStore with some initial info.
+        // 2. Is cookieStoreId null? This means the caller probably wants everything currently in the proxifiedContainersStore object store
+        // 3. If there doesn't exist an entry for the associated cookieStoreId, inform the caller of this
+        // 4. Normal operation - if the cookieStoreId exists in the map, we can simply resolve with the correct proxy value
+
+        const results_array = results["proxifiedContainersKey"];
+
+        if (Object.getOwnPropertyNames(results).length === 0) {
+          reject({
+            error: "uninitialized",
+            message: ""
+          });
+        } else if (cookieStoreId === null) {
+          resolve(results_array);
+        } else {
+          const val = results_array.find(o => o.cookieStoreId === cookieStoreId);
+
+          if (typeof val !== "object" || val === null) {
+            reject({
+              error: "doesnotexist",
+              message: ""
+            });
+          } else {
+            resolve(val);
+          }
+        }
+
+      }, (error) => {
+        reject({
+          error: "internal",
+          message: error
+        });
+      }).catch((error) => {
+        proxifiedContainers.report_proxy_error(error, "proxified-containers.js: error 1");
+      });
+    });
+  },
+  set: function(cookieStoreId, proxy, initialize = false) {
+    return new Promise((resolve, reject) => {
+      if (initialize === true) {
+        const proxifiedContainersStore = [];
+        proxifiedContainersStore.push({
+          cookieStoreId: cookieStoreId,
+          proxy: proxy
+        });
+
+        browser.storage.local.set({
+          proxifiedContainersKey: proxifiedContainersStore
+        });
+
+        resolve(proxy);
+      }
+
+      // Assumes proxy is a properly formatted object
+      proxifiedContainers.retrieve().then((proxifiedContainersStore) => {
+        let index = proxifiedContainersStore.findIndex(i => i.cookieStoreId === cookieStoreId);
+
+        if (index === -1) {
+          proxifiedContainersStore.push({
+            cookieStoreId: cookieStoreId,
+            proxy: proxy
+          });
+          index = proxifiedContainersStore.length - 1;
+        } else {
+          proxifiedContainersStore[index] = {
+            cookieStoreId: cookieStoreId,
+            proxy: proxy
+          };
+        }
+
+        browser.storage.local.set({
+          proxifiedContainersKey: proxifiedContainersStore
+        });
+
+        resolve(proxifiedContainersStore[index]);
+      }, (errorObj) => {
+        reject(errorObj);
+      }).catch((error) => {
+        throw error;
+      });
+    });
+  },
+  parseProxy: function(proxy_str) {
+    const regexp = /(\b(\w+):(\w+)@)?(((?:\d{1,3}\.){3}\d{1,3}\b)|(\b(\w+)(\.(\w+))+))(:(\d+))?/;
```

Comment:
Add a comment explaining what this is parsing so we don't forget please :)

Reply:
Should be resolved now. Thanks